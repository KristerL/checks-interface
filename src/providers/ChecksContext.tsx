import { createContext } from 'react';

const ValidKeyboardKeys = ['ArrowUp', 'ArrowDown', '1', '2'] as const;
export const isValidKeyboardKey = (value: any): value is KeyboardKeys => ValidKeyboardKeys.includes(value);
export type KeyboardKeys = typeof ValidKeyboardKeys[number];

export type CheckButtonValues = 'yes' | 'no' | null;

export interface Check {
  id: string;
  priority: number;
  description: string;
  value: CheckButtonValues;
}

export interface ChecksContextProps {
  checks: { [key: string]: Check };
  activeCheckIndex: number;
  lastValidIndex: number;
  loadChecks: () => Promise<any>;
  changeCheckValue: (key: string, newValue: Exclude<CheckButtonValues, null>, index: number) => void;
  handleKeyBoardEvent: (key: KeyboardKeys) => void;
}

export const initialState: ChecksContextProps = {
  checks: {},
  activeCheckIndex: null,
  lastValidIndex: 0,

  loadChecks: () => null,
  changeCheckValue: () => null,
  handleKeyBoardEvent: () => null,
};

export const ChecksContext = createContext<ChecksContextProps>(initialState);
