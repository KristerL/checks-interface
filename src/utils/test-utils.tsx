import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import {
  ChecksContextProvider,
  ChecksContextProviderProps,
  ChecksContextState,
} from '../providers/ChecksContextProvider';

const AllTheProviders: FC<ChecksContextProviderProps> = ({ children, customInitialState }) => {
  return <ChecksContextProvider customInitialState={customInitialState}>{children}</ChecksContextProvider>;
};

interface CustomRenderProps extends RenderOptions {
  initialState?: ChecksContextState;
}

const customRender = (ui: ReactElement, customRenderProps?: CustomRenderProps) => {
  const { ...renderProps } = customRenderProps || {};

  return render(ui, {
    wrapper: (props) => (
      <AllTheProviders customInitialState={customRenderProps?.initialState}>{props.children}</AllTheProviders>
    ),
    ...renderProps,
  });
};

export * from '@testing-library/react';
export { customRender as render };
