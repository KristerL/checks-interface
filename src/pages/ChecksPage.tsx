import React, { useContext, useEffect, useState } from 'react';
import { ChecksView, SubmitData } from '../components/checksView/ChecksView';
import { ChecksContext } from '../providers/ChecksContext';
import { submitCheckResults } from '../api';
import Button from '../components/button/Button';

type errorType = {
  message: string;
  type: 'submit' | 'load';
};

export const ChecksPage = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<errorType>(null);
  const [submitted, setSubmitted] = useState(false);
  const { loadChecks } = useContext(ChecksContext);

  useEffect(() => {
    handleCheckLoad();
  }, []);

  const handleCheckLoad = () => {
    loadChecks()
      .catch((err: errorType) => {
        setError(err);
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  const handleSubmit = (data: SubmitData) => {
    setIsLoading(true);
    submitCheckResults(data)
      .then(() => {
        setError(null);
        setSubmitted(true);
      })
      .catch((err) => setError(err))
      .finally(() => {
        setIsLoading(false);
      });
  };

  const renderErrorMessage = () => {
    const reloadData = () => {
      setIsLoading(true);
      setError(null);
      handleCheckLoad();
    };

    return (
      <div className="errorMessageContainer">
        <h1 className="centerText">{error.message}</h1>
        {error.type === 'load' && (
          <Button classname="submitButton" handleOnClick={reloadData}>
            Try to load again
          </Button>
        )}
      </div>
    );
  };

  const refreshPage = () => {
    window.location.reload();
  };

  const renderLoadingView = () => {
    return (
      <div className="viewContainer">
        <h1 className="centerText">Loading...</h1>
      </div>
    );
  };

  const renderSubmitView = () => {
    return (
      <div className="viewContainer">
        <h1>Submit was successful</h1>
        <Button classname="submitButton" handleOnClick={refreshPage}>
          Start again
        </Button>
      </div>
    );
  };

  if (submitted) {
    return renderSubmitView();
  }

  if (isLoading) {
    return renderLoadingView();
  }

  return (
    <div className="viewContainer">
      {error && renderErrorMessage()}
      <ChecksView handleSubmit={handleSubmit} />
    </div>
  );
};
