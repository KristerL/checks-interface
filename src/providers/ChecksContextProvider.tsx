import React, { useState, useMemo } from 'react';
import { Check, CheckButtonValues, ChecksContext, KeyboardKeys } from './ChecksContext';
import { fetchChecks } from '../api';

interface ChecksContextState {
  checks: { [key: string]: Check };
  activeCheckIndex: number;
}

const initialState: ChecksContextState = {
  checks: {},
  activeCheckIndex: null,
};

export const ChecksContextProvider: React.FC = ({ children }) => {
  const [state, setState] = useState(initialState);

  const loadChecks = () => {
    return fetchChecks()
      .then((res) => {
        const sortedChecks = res.sort((a, b) => b.priority - a.priority);
        const checksObject = sortedChecks.reduce((acc, check) => {
          acc[check.id] = check;
          return acc;
        }, {});
        setState({ ...state, checks: checksObject });
        return Promise.resolve();
      })
      .catch((err) => {
        return Promise.reject(err);
      });
  };

  const createNewCheckObject = (key: string, newValue: CheckButtonValues, state: ChecksContextState) => {
    return {
      ...state.checks,
      [key]: {
        ...state.checks[key],
        value: newValue,
      },
    };
  };

  const changeCheckValue = (key: string, newValue: CheckButtonValues, index: number) => {
    const newCheck = createNewCheckObject(key, newValue, state);
    setState({ ...state, checks: newCheck, activeCheckIndex: index });
  };

  const handleKeyboardNavigation = (key: 'ArrowUp' | 'ArrowDown', prevState: ChecksContextState) => {
    const { checks, activeCheckIndex } = prevState;
    const selectedCheckIndex = activeCheckIndex === null ? -1 : activeCheckIndex;

    const addition = key === 'ArrowUp' ? -1 : 1;
    const newCheckIndex = selectedCheckIndex + addition;

    const isNewIndexOutOfBounds = newCheckIndex < 0 || newCheckIndex >= Object.keys(checks).length;
    const isNewIndexDisabled = newCheckIndex > Object.values(checks).filter(check => check.value).length;

    if (isNewIndexOutOfBounds || isNewIndexDisabled) {
      return prevState;
    }

    return {
      ...prevState,
      activeCheckIndex: newCheckIndex,
    };
  };

  const handleKeyboardSelect = (key: '1' | '2', prevState: ChecksContextState) => {
    const { checks, activeCheckIndex } = prevState;

    if (activeCheckIndex === null) {
      return prevState;
    }

    const currentCheck = checks[Object.keys(checks)[activeCheckIndex]];
    const newValue: CheckButtonValues = key === '1' ? 'yes' : 'no';
    const newCheck = createNewCheckObject(currentCheck.id, newValue, prevState);

    return {
      ...prevState,
      checks: newCheck,
      activeCheckIndex,
    };
  };

  const handleKeyBoardEvent = (key: KeyboardKeys) => {
    setState((prevState) => {
      let newState: ChecksContextState;
      if (key === '1' || key === '2') {
        newState = handleKeyboardSelect(key, prevState);
      } else {
        newState = handleKeyboardNavigation(key, prevState);
      }
      return newState;
    });
  };

  const contextValue = useMemo(
    () => ({
      ...state,
      loadChecks,
      changeCheckValue,
      handleKeyBoardEvent,
    }),
    [state]
  );

  return <ChecksContext.Provider value={contextValue}>{children}</ChecksContext.Provider>;
};
