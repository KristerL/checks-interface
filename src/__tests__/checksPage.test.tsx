import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react';
import { render } from '../utils/test-utils';
import { checks, firstRowDescription, singleCheck } from './constants';
import { ChecksContextState } from '../providers/ChecksContextProvider';
import { ChecksPage } from '../pages/ChecksPage';
import * as apiCalls from '../api';

const renderCheckPage = () => {
  const customInitialState: ChecksContextState = {
    checks: checks,
    activeCheckIndex: null,
    lastValidIndex: 0,
  };

  return render(<ChecksPage />, { initialState: customInitialState });
};

const mockedFetchCall = jest.spyOn(apiCalls, 'fetchChecks');
const mockedSubmitCall = jest.spyOn(apiCalls, 'submitCheckResults');

beforeEach(() => {
  // resolve is the default course of action
  mockedFetchCall.mockImplementation(() => new Promise((resolve) => resolve([singleCheck])));
  mockedSubmitCall.mockImplementation((result) => new Promise((resolve) => resolve(result)));
});
afterEach(() => {
  mockedFetchCall.mockRestore();
  mockedSubmitCall.mockRestore();
});

describe('checksPage tests', () => {
  it('should match snapshot', async function () {
    const { container, findByText } = renderCheckPage();

    expect(await findByText(firstRowDescription));
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('should initially show loading screen', async () => {
    const { getByText, findByText } = renderCheckPage();

    expect(getByText('Loading...'));
    expect(await findByText(firstRowDescription));
  });

  it('should notify user when fetch failed', async () => {
    const mockedCall = jest.spyOn(apiCalls, 'fetchChecks');
    mockedCall.mockImplementation(
      () =>
        new Promise((resolve, reject) =>
          reject({
            type: 'load',
            message: 'No checks were found to display',
          })
        )
    );

    const { findByText, getByText } = renderCheckPage();

    expect(getByText('Loading...'));
    expect(await findByText('No checks were found to display'));
  });

  it('should be able to fetch data again', async function () {
    const mockedCall = jest.spyOn(apiCalls, 'fetchChecks');
    mockedCall.mockImplementation(
      () =>
        new Promise((resolve, reject) =>
          reject({
            type: 'load',
            message: 'No checks were found to display',
          })
        )
    );

    const { findByText, getByText } = renderCheckPage();

    expect(getByText('Loading...'));
    expect(await findByText('No checks were found to display'));

    mockedCall.mockImplementation(() => new Promise((resolve) => resolve([singleCheck])));

    const retryButton = getByText('Try to load again');
    fireEvent.click(retryButton);

    expect(getByText('Loading...'));
    expect(await findByText(firstRowDescription));
  });

  it('should notify user when no checks were retrieved', async () => {
    const mockedFetchCall = jest.spyOn(apiCalls, 'fetchChecks');
    mockedFetchCall.mockImplementation(() => new Promise((resolve) => resolve([])));
    const { findByText } = renderCheckPage();

    expect(await findByText('No checks were found to display'));
  });

  it('should say when data is submitted', async () => {
    const mockedFetchCall = jest.spyOn(apiCalls, 'fetchChecks');
    mockedFetchCall.mockImplementation(() => new Promise((resolve) => resolve([singleCheck])));
    const { findByText, getByText } = renderCheckPage();

    expect(await findByText(firstRowDescription));
    const yesButton = getByText('yes');
    const submitButton = getByText('SUBMIT');

    fireEvent.click(yesButton);
    fireEvent.click(submitButton);

    expect(await findByText('Submit was successful'));
  });

  it('should say when data submit failed and retry', async () => {
    const mockedFetchCall = jest.spyOn(apiCalls, 'fetchChecks');
    mockedFetchCall.mockImplementation(() => new Promise((resolve) => resolve([singleCheck])));
    const mockedSubmitCall = jest.spyOn(apiCalls, 'submitCheckResults');
    mockedSubmitCall.mockImplementation(
      () =>
        new Promise((resolve, reject) =>
          reject({
            type: 'submit',
            message: 'Something went wrong while submitting. Please try again',
          })
        )
    );
    const { findByText, getByText } = renderCheckPage();

    expect(await findByText(firstRowDescription));
    const yesButton = getByText('yes');
    let submitButton = getByText('SUBMIT');

    fireEvent.click(yesButton);
    fireEvent.click(submitButton);

    expect(await findByText('Something went wrong while submitting. Please try again'));

    mockedSubmitCall.mockImplementation((result) => new Promise((resolve) => resolve(result)));

    submitButton = getByText('SUBMIT');
    fireEvent.click(submitButton);

    expect(await findByText('Submit was successful'));
  });
});
