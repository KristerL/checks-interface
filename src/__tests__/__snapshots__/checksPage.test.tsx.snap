// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checksPage tests should match snapshot 1`] = `
<div>
  <div
    class="viewContainer"
  >
    <div>
      <div
        class="checkRow"
      >
        <h1
          class="checkDescription"
        >
          Veriff supports presented document
        </h1>
        <div
          class="buttonContainer"
        >
          <button
            class="button toggleButton"
          >
            yes
          </button>
          <button
            class="button toggleButton"
          >
            no
          </button>
        </div>
      </div>
      <div
        class="submitButtonContainer"
      >
        <button
          class="button disabled submitButton"
          disabled=""
        >
          SUBMIT
        </button>
      </div>
    </div>
  </div>
</div>
`;
